compiler/resolution/cleanups.cpp

```diff
@@ -885,8 +894,69 @@ static void cleanupVoidVarsAndFields() {
   }
 }
 
+void saveGenericSubstitutions() {
+  for_alive_in_Vec(FnSymbol, fn, gFnSymbols) {
+    if (fn->substitutions.n > 0) {
+
+      fn->substitutionsPostResolve.clear();
+
+      if (FnSymbol* genericFn = toFnSymbol(fn->instantiatedFrom)) {
+        // Construct substitutionsPostResolve
+        for_formals(genericFormal, genericFn) {
+          if (genericFormal->type != dtMethodToken) {
+            Symbol* value = fn->substitutions.get(genericFormal);
+            NameAndSymbol ns;
+            ns.name = genericFormal->name;
+            ns.value = value;
+            ns.isParam = genericFormal->isParameter();
+            ns.isType = (genericFormal->originalIntent == INTENT_TYPE);
+            fn->substitutionsPostResolve.push_back(ns);
+          }
+        }
+
+      } else {
+        // This case is a workaround for patterns that
+        // come up with compiler-generated tuple functions
+        INT_ASSERT(fn->hasFlag(FLAG_INIT_TUPLE));
+        form_Map(SymbolMapElem, e, fn->substitutions) {
+          NameAndSymbol ns;
+          ns.name = e->key->name;
+          ns.value = e->value;
+          ns.isParam = false;
+          ns.isType = false;
+          fn->substitutionsPostResolve.push_back(ns);
+        }
+      }
+
+      // Clear substitutions since keys might refer to deleted AST nodes
+      fn->substitutions.clear();
+    }
+  }
+
+  for_alive_in_Vec(TypeSymbol, ts, gTypeSymbols) {
+    if (AggregateType* at = toAggregateType(ts->type)) {
+      if (at->substitutions.n > 0) {
+        // Construct substitutionsPostResolve
+        at->substitutionsPostResolve.clear();
```

Comment:
Why's this necessary? Or can we check whether this is populated and `continue` if it is?

Reply:
It should not be already populated. I added this out of some desire to create robust code. I could alternatively assert that it is empty if you like that better.

Reply:
I've changed it to an assert.